File: problems_026to050/problem_043/solution.py
import os
import sys
from collections import Counter

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch
from toolbox import smaller_primes


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = 0

        pandigital_set = Counter('0123456789')

        def suffixes_with_divisibility_property(prefix):
            primes = smaller_primes(18)

            def recursive_closure(prefix, primes):
                if len(primes) == 0:
                    return None

                result = []
                prime = primes[0]

                for subnumber in range(0, 1000, prime):
                    if prefix % 100 != subnumber // 10:
                        continue

                    units_place = subnumber % 10
                    longer_prefix = str(prefix) + str(units_place)

                    if len(Counter(longer_prefix) - pandigital_set) > 0:
                        continue

                    suffixes = recursive_closure(int(longer_prefix), primes[1:])

                    if suffixes is None:
                        result.append(str(units_place))
                    else:
                        result += [str(units_place) + suffix for suffix in suffixes]

                return result

            return recursive_closure(prefix, primes)

        for prefix in range(102, 987+1):
            suffixes = suffixes_with_divisibility_property(prefix)

            for suffix in suffixes:
                decimal = str(prefix) + suffix

                if len(Counter(decimal) - pandigital_set) == 0 and len(pandigital_set - Counter(decimal)) == 0:
                    answer += int(decimal)

        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_026to050/problem_043/solution.js
const rootPath = require('app-root-path');
const { setDifference } = require('mathjs');

const Stopwatch = require(`${rootPath}/lib/Stopwatch.js`);

class Solution extends Stopwatch {
  execute() {
    let answer = 0;

    const pandigitalSet = '0123456789'.split('');

    const suffixesWithDivisibilityProperty = (initialPrefix) => {
      const recursion = (prefix, primes) => {
        if (primes.length === 0) {
          return null;
        }

        const result = [];
        const [prime] = primes;

        for (let subnumber = 0; subnumber < 1000; subnumber += prime) {
          if (prefix % 100 === Math.floor(subnumber / 10)) {
            const unitsPlace = subnumber % 10;
            const longerPrefix = String(prefix) + String(unitsPlace);

            if (setDifference(longerPrefix.split(''), pandigitalSet).length === 0) {
              const suffixes = recursion(Number(longerPrefix), primes.slice(1));

              if (suffixes === null) {
                result.push(String(unitsPlace));
              }
              else {
                result.push(...suffixes.map((suffix) => String(unitsPlace) + suffix));
              }
            }
          }
        }

        return result;
      };

      return recursion(initialPrefix, [2, 3, 5, 7, 11, 13, 17]);
    };

    for (let prefix = 102; prefix <= 987; prefix += 1) {
      const suffixes = suffixesWithDivisibilityProperty(prefix);

      for (let index = 0; index < suffixes.length; index += 1) {
        const suffix = suffixes[index];
        const decimal = String(prefix) + suffix;

        if (setDifference(decimal.split(''), pandigitalSet).length === 0 && setDifference(pandigitalSet, decimal.split('')).length === 0) {
          answer += Number(decimal);
        }
      }
    }

    return answer;
  }
}

(() => {
  const solution = new Solution();

  const result = solution.execute();

  // eslint-disable-next-line no-console
  console.log(result);
})();
